File: problem_034.py
#!/usr/bin/env python

print("Project Euler - Problem 34:")
print("Find the sum of all the numbers that can be written as the sum of fifth powers of their digits\n")

def factorial(n):
    if n == 0:
        return 1
    else:
        return n * factorial(n-1)

dict={"0":factorial(0), "1":factorial(1), "2":factorial(2),
      "3":factorial(3), "4":factorial(4), "5":factorial(5),
      "6":factorial(6), "7":factorial(7), "8":factorial(8),
      "9":factorial(9)}

answer = 0
limit = 100000
for n in range(10, limit):
    s = 0
    for d in str(n):
        s += dict[d]
    if n == s:
        answer += n

print(answer)

File: problem_034.js
#!/usr/bin/node

console.log("Project Euler - Problem 34:");
console.log("Find the sum of all the numbers that can be written as the sum of the factorials of their digits\n");

function is_factorial_sum(n, factorial){
    let t = n;
    var sum = 0;

    // Decompose number and create factorial sum
    while (t > 0 && sum <= n){
        sum += factorial[Math.floor(t % 10)];
        t = Math.floor(t / 10);
    }

    // Compare sum
    if (sum == n)
        return true;

    return false;
}

// Pre-populate factorials
var factorials = [1, 1, 1, 1, 1, 1, 1, 1, 1, 1];
for (var i=1; i<10; i++)
    for (var n=i; n<10; n++)
        factorials[n] *= i;

var sum = 0;
const limit = 100000;

for (var n=10; n<limit; n++)
    if (is_factorial_sum(n, factorials))
        sum += n;

console.log(sum);
